blas/impl/KokkosBlas2_ger_impl.hpp

```diff
@@ -0,0 +1,339 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOSBLAS2_GER_IMPL_HPP_
+#define KOKKOSBLAS2_GER_IMPL_HPP_
+
+#include "KokkosKernels_config.h"
+#include "Kokkos_Core.hpp"
+#include "KokkosKernels_ExecSpaceUtils.hpp"
+#include "Kokkos_ArithTraits.hpp"
+
+namespace KokkosBlas {
+namespace Impl {
+
+// Functor for a single-level parallel_for version of nontranspose GER.
+// The functor parallelizes over rows of the input matrix A.
+template <class XViewType, class YViewType, class AViewType,
+          class IndexType = typename AViewType::size_type>
+struct SingleLevelGER {
+  using AlphaCoeffType = typename AViewType::non_const_value_type;
+  using A_value_type   = typename AViewType::non_const_value_type;
+
+  SingleLevelGER( const bool             justTranspose
+                , const AlphaCoeffType & alpha
+                , const XViewType      & x
+                , const YViewType      & y
+                , const AViewType      & A
+                )
+      : justTranspose_(justTranspose)
+      , alpha_        (alpha)
+      , x_            (x)
+      , y_            (y)
+      , A_            (A)
+  {
+    static_assert(Kokkos::is_view<XViewType>::value, "XViewType must be a Kokkos::View.");
+    static_assert(Kokkos::is_view<YViewType>::value, "YViewType must be a Kokkos::View.");
+    static_assert(Kokkos::is_view<AViewType>::value, "AViewType must be a Kokkos::View.");
+
+    static_assert(static_cast<int>(XViewType::rank) == 1, "XViewType must have rank 1.");
+    static_assert(static_cast<int>(YViewType::rank) == 1, "YViewType must have rank 1.");
+    static_assert(static_cast<int>(AViewType::rank) == 2, "AViewType must have rank 2.");
+
+    static_assert(std::is_integral<IndexType>::value, "IndexType must be an integer.");
+  }
+
+  KOKKOS_INLINE_FUNCTION void operator()(const IndexType & i) const {
+    using KAT = Kokkos::ArithTraits<typename AViewType::non_const_value_type>;
+
+    if (alpha_ == KAT::zero()) {
+      // Nothing to do
+    }
+    else {
+      const IndexType    N      ( A_.extent(1) );
+      const A_value_type x_fixed( x_(i) );
+
+      if (justTranspose_) {
+        for (IndexType j = 0; j < N; ++j) {
+          A_(i,j) += A_value_type( alpha_ * x_fixed * y_(j) );
+        }
+      }
+      else {
+        for (IndexType j = 0; j < N; ++j) {
+          A_(i,j) += A_value_type( alpha_ * x_fixed * KAT::conj( y_(j) ) );
```

Comment:
I am not sure if `KAT::conj( y_(j) )` is valid for every possible type combination that we will face.
`KAT = Kokkos::ArithTraits<typename AViewType::non_const_value_type>` uses the `non_const_value_type` of `A` not of `y`.
To be on the safe side, I would suggest to use two `Kokkos::ArithTraits` one with the `non_const_value_type` of `A` when it is used for entries of `A` and one with the `non_const_value_type` of `y` when it is used for entries of `y`.

Reply:
Done. Thanks for the hint.